src/cargo/sources/config.rs

```diff
@@ -0,0 +1,156 @@
+    replace_with: Option<(String, PathBuf)>,
+}
+
+impl<'cfg> SourceConfigMap<'cfg> {
+    pub fn new(config: &'cfg Config) -> CargoResult<SourceConfigMap<'cfg>> {
+        let mut base = try!(SourceConfigMap::empty(config));
+        if let Some((table, _path)) = try!(config.get_table("source")) {
+            for (key, value) in table.iter() {
+                try!(base.add_config(key, value));
+            }
+        }
+        Ok(base)
+    }
+
+    pub fn empty(config: &'cfg Config) -> CargoResult<SourceConfigMap<'cfg>> {
+        let mut base = SourceConfigMap {
+            cfgs: HashMap::new(),
+            id2name: HashMap::new(),
+            config: config,
+        };
+        base.add("crates-io", SourceConfig {
+            id: try!(SourceId::crates_io(config)),
+            replace_with: None,
+        });
+        Ok(base)
+    }
+
+    pub fn load(&self, id: &SourceId) -> CargoResult<Box<Source + 'cfg>> {
+        debug!("loading: {}", id);
+        let mut name = match self.id2name.get(id) {
+            Some(name) => name,
+            None => return Ok(id.load(self.config)),
+        };
+        let mut path = Path::new("/");
+        let orig_name = name;
+        let new_id;
+        loop {
+            let cfg = match self.cfgs.get(name) {
+                Some(cfg) => cfg,
+                None => bail!("could not find a configured source with the \
+                               name `{}` when attempting to lookup `{}` \
+                               (configuration in `{}`)",
+                              name, orig_name, path.display()),
+            };
+            match cfg.replace_with {
+                Some((ref s, ref p)) => {
+                    name = s;
+                    path = p;
+                }
+                None if *id == cfg.id => return Ok(id.load(self.config)),
+                None => {
+                    new_id = cfg.id.with_precise(id.precise()
+                                                 .map(|s| s.to_string()));
+                    break
+                }
+            }
+            debug!("following pointer to {}", name);
+            if name == orig_name {
+                bail!("detected a cycle of `replace-with` sources, the source \
+                       `{}` is eventually replaced with itself \
+                       (configuration in `{}`)", name, path.display())
+            }
+        }
+        let new_src = new_id.load(self.config);
+        let old_src = id.load(self.config);
+        if new_src.supports_checksums() != old_src.supports_checksums() {
+            let (supports, no_support) = if new_src.supports_checksums() {
+                (name, orig_name)
+            } else {
+                (orig_name, name)
+            };
+            bail!("\
+cannot replace `{}` with `{}`, the source `{}` supports checksums,
+but `{}` does not
+
+a lock file compatible with `{}` cannot be generated in this situation
+", orig_name, name, supports, no_support, orig_name);
```

Comment:
Five substitutions are hard to read, I'd try to use named params (or however `bail!("{orig}", orig=orig_name)` is called)